rllib/examples/rl_modules/classes/vpg_using_shared_encoder_rlm.py

```diff
@@ -0,0 +1,164 @@
+        with torch.no_grad():
+            return self._common_forward(batch)
+
+    @override(RLModule)
+    def _forward_train(self, batch):
+        return self._common_forward(batch)
+
+    def _common_forward(self, batch):
+        # Features can be found in the batch under the "encoder_features" key.
+        features = batch["encoder_features"]
+        logits = self.policy_head(features)
+        return {Columns.ACTION_DIST_INPUTS: logits}
+
+
+class VPGTorchMultiRLModuleWithSharedEncoder(MultiRLModule):
+    """A VPG (vanilla policy gradient)-style MultiRLModule with shared encoder.
+
+    This MultiRLModule needs to be configured appropriately as follows:
+
+    .. testcode::
+        :skipif: true
+
+        from ray.rllib.core.rl_module.multi_rl_module import MultiRLModuleSpec
+        from ray.rllib.core.rl_module.rl_module import RLModuleSpec
+
+        FEATURE_DIM = 64  # encoder output (feature) dim
+        HIDDEN_DIM = 64  # hidden dim for the policy nets
+
+        config.rl_module(
+            rl_module_spec=MultiRLModuleSpec(
+                module_specs={
+                    # Central/shared encoder net.
+                    SHARED_ENCODER_ID: RLModuleSpec(
+                        module_class=SharedTorchEncoder,
+                        model_config_dict={"feature_dim": FEATURE_DIM},
+                    ),
+                    # Arbitrary number of policy nets (w/o encoder sub-net).
+                    "p0": RLModuleSpec(
+                        module_class=VPGTorchRLModuleUsingSharedEncoder,
+                        model_config_dict={
+                            "feature_dim": FEATURE_DIM,
+                            "hidden_dim": HIDDEN_DIM,
+                        },
+                    ),
+                    "p1": RLModuleSpec(
+                        module_class=VPGTorchRLModuleUsingSharedEncoder,
+                        model_config_dict={
+                            "feature_dim": FEATURE_DIM,
+                            "hidden_dim": HIDDEN_DIM,
+                        },
+                    ),
+                },
+            ),
+        )
+
+    Also note that in order to learn properly, a special, multi-agent Learner that
+    accounts for the shared encoder must be setup. This Learner should have only a
+    single optimizer (for all submodules: encoder and all policy nets) in order to not
+    destabilize learning. The latter would happen if more than one optimizer would try
+    to optimize the same shared encoder submodule.
+    """
+
+    @override(MultiRLModule)
+    def setup(self):
+        super().setup()
+
+        # Assert, we have the shared encoder submodule.
+        assert (
+            SHARED_ENCODER_ID in self._rl_modules
+            and isinstance(self._rl_modules[SHARED_ENCODER_ID], SharedTorchEncoder)
+            and len(self._rl_modules) > 1
+        )
+
+    @override(MultiRLModule)
+    def _run_forward_pass(self, forward_fn_name, batch, **kwargs):
```

Comment:
This is still a strange form - passing in the `forward_fn_name`. Why is this needed? To have a single method to be overridden in inheritance?

Reply:
This will be gone in another PR that's already in (your) review :)

Reply:
Great question!